### PR TITLE
T275621 Experiment Config

### PR DESCRIPTION
**Pull request by hueitan**
Phabricator Link: https://phabricator.wikimedia.org/T275621

### Problem Statement

The app should query the config and turn the experiment on or off accordingly.

### Solution

Experiment Configuration getting from the config page

- [x] Get API
- [x] Setting experiment config to allow start/end/countries/languages
- [x] read the user region from xhr
- [x] Don't get the new config if the config existed less than 1 day
 
### Note

Usage

```js
import { useExperimentConfig } from 'hooks'
const isExperimentGroup = useExperimentConfig(lang)
```

### CODE REVIEW

File: src/hooks/useExperimentConfig.js
Comment (stephanebisson) [inline src/hooks/useExperimentConfig.js:28]: Please keep `xhr` encapsulated inside cachedFetch. You can extract and parse the cookies in the API layer.

File: src/hooks/useExperimentConfig.js
Comment (medied) [inline src/hooks/useExperimentConfig.js:12]: Use `let`/`const` instead of `var`?

**Comment by hueitan** [comment]
Thanks for the great review, it's easier to read now and ready for review @stephanebisson @medied

**Comment by hueitan** [comment]
now the branch is merged with master, if you want to see the trending feed, do remember changing the local storage `user-country` to `NG`

File: src/hooks/useExperimentConfig.js
Comment (medied) [inline src/hooks/useExperimentConfig.js:21]: I appreciate this sort of naming

File: src/hooks/useExperimentConfig.js
Comment (medied) [inline src/hooks/useExperimentConfig.js:51]: In the experiment context, how does a record expire exactly?

**Comment by medied** [comment]
@hueitan looking good, well done. I'd like to review it again after the conflicts have been resolved

File: src/hooks/useExperimentConfig.js
Comment (hueitan) [reply src/hooks/useExperimentConfig.js:51]: the record checks on line 44, expire after 1 day.

1 day is my assumption, we can change it if think too much or less.

**Comment by hueitan** [comment]
> @hueitan looking good, well done. I'd like to review it again after the conflicts have been resolved

Thanks @medied I will wait until the revert revert pr merge then merge again master to this branch.

File: src/utils/cachedFetch.js
Comment (medied) [inline src/utils/cachedFetch.js:35]: We have `USER_COUNTRY_STORAGE_KEY` in `useExperimentConfig` and here we are setting 'user-country' directly. Honestly I think it's fine this way... but thought I'd flag it just in case. If we ever need to change this key we have to remember to change it in both places.

**Review by medied — APPROVED** [review]

File: src/hooks/useExperimentConfig.js
Comment (stephanebisson) [inline src/hooks/useExperimentConfig.js:6]: I would like to see user country get/set be encapsulated in a util.

**Review by stephanebisson — CHANGES_REQUESTED** [review]
One small change requested.

This is kinda hard to test. Would it be crazy to put some controls in the simulator to set user country, date, experiment group, etc?

**Comment by jpita** [comment]
> One small change requested.
> 
> This is kinda hard to test. Would it be crazy to put some controls in the simulator to set user country, date, experiment group, etc?

this would be super handy

File: sim.html
Comment (hueitan) [reply sim.html]: Yes, tested, this was removed by mistake, and already update in the next commit.

File: src/hooks/useExperimentConfig.js
Comment (stephanebisson) [reply src/hooks/useExperimentConfig.js]: I still think user country get/set should be a util

**Comment by stephanebisson** [comment]
Can you make the control panel always visible?

File: sim.html
Comment (hueitan) [inline sim.html:77]: hmm, i should make a record whether on/off the next time user visit the simulator. working on it.

File: sim.html
Comment (stephanebisson) [inline sim.html:184]: It overlaps the device when I have the dev tools open. You can let it flow below the device or put it in its own box on the side but no fixed or absolute position.

File: sim.html
Comment (stephanebisson) [inline sim.html:424]: Not sure about this condition. When I load the sim this box is checked because I have a user-country setup but I don't meet all the other conditions for the experiment to be ON. It could be unchecked by default. It would not be perfect but it would be much better.

**Review by stephanebisson — CHANGES_REQUESTED** [review]
Looks good. Just some final details.

**Pull request merged**
